kyo-kernel/shared/src/main/scala/kyo/kernel/Isolate.scala

```diff
@@ -0,0 +1,370 @@
+
+        /** Phase 3: State Restoration
+          *
+          * Restores/merges state after isolated execution completes. Determines how transformed state is propagated back to the outer
+          * context.
+          */
+        def restore[A: Flat, S](v: Transform[A] < S)(using Frame): A < (S & Retain & Passthrough)
+
+        /** Applies this isolate to a computation that requires it.
+          *
+          * Provides a more ergonomic way to use isolates with effects:
+          * {{{
+          * Var.isolate.update[Int].use {
+          *   Async.mask {
+          *     // computation with isolated Var[Int] effect
+          *   }
+          * }
+          * }}}
+          *
+          * @param f
+          *   The computation requiring this isolate
+          * @return
+          *   The result of running the computation with this isolate
+          */
+        def use[A](f: this.type ?=> A): A = f(using this)
+
+        /** Composes this isolate with another, managing both states.
+          *
+          * Creates a new isolate that handles the state lifecycles of both this isolate and the next one, maintaining proper ordering and
+          * effect tracking.
+          */
+        def andThen[R2, P2](next: Stateful[R2, P2]): Stateful[Retain & R2, Passthrough & P2] =
+            new Stateful[Retain & R2, Passthrough & P2]:
+                type State        = (self.State, next.State)
+                type Transform[A] = self.Transform[next.Transform[A]]
+                given flatTransform[A: Flat]: Flat[Transform[A]] = self.flatTransform(using next.flatTransform)
+
+                def capture[A: Flat, S2](f: State => A < S2)(using Frame) =
+                    self.capture(s1 => next.capture(s2 => f((s1, s2))))
+
+                def isolate[A: Flat, S3](state: (self.State, next.State), v: A < (Retain & R2 & S3))(using Frame) =
+                    self.isolate(state._1, next.isolate(state._2, v))(using Flat.unsafe.bypass)
+
+                def restore[A: Flat, S2](v: self.Transform[next.Transform[A]] < S2)(using Frame) =
+                    next.restore(self.restore(v)(using Flat.unsafe.bypass))
+            end new
+        end andThen
+    end Stateful
+
+    object Stateful:
+
+        /** No-op isolate with no state. */
+        val noop: Stateful[Any, Any] =
+            new Stateful[Any, Any]:
+                type State        = Unit
+                type Transform[A] = A
+                given flatTransform[A: Flat]: Flat[A]                                 = Flat.derive[A]
+                def capture[A: Flat, S2](f: State => A < S2)(using Frame)             = f(())
+                def isolate[A: Flat, S2](state: Unit, v: A < (Any & S2))(using Frame) = v
+                def restore[A: Flat, S2](v: A < S2)(using Frame)                      = v
+                override def andThen[R2, P2](next: Stateful[R2, P2])                  = next
```

Comment:
The new encoding makes `noop` much more efficient since it avoids allocations in all its methods.